hal/armv8m/nrf/91/console.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * HAL console (nRF9160 UART)
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "nrf91.h"
+
+#include "../../../console.h"
+#include "../../../cpu.h"
+#include "../../../string.h"
+#include "../../armv8m.h"
+#include <board_config.h>
+
+
+#define TX_DMA_SIZE 64
+
+
+static struct {
+	volatile u32 *base;
+	u8 txPin;
+	u8 rxPin;
+	u8 rtsPin;
+	u8 ctsPin;
+	char txDma[TX_DMA_SIZE];
+	int txDmaSize;
+} console_common;
+
+
+/* clang-format off */
+enum { uarte_startrx = 0, uarte_stoprx, uarte_starttx, uarte_stoptx, 
+	uarte_events_cts = 64, uarte_events_txdrdy = 71, uarte_events_endtx, uarte_events_error, uarte_events_txstarted = 84, 
+	uarte_inten = 192, uarte_errorsrc = 288, uarte_intenset, uarte_intenclr, uarte_enable = 320, 
+	uarte_psel_rts = 322, uarte_psel_txd, uarte_psel_cts, uarte_psel_rxd, uarte_baudrate = 329, 
+	uarte_rxd_ptr = 333, uarte_rxd_maxcnt, uarte_rxd_amount, uarte_txd_ptr = 337, uarte_txd_maxcnt, uarte_txd_amount, 
+	uarte_config = 347 };
+/* clang-format on */
+
+
+/* Init pins according to nrf9160 product specification */
+static void console_configPins(void)
+{
+	_nrf91_gpioConfig(console_common.txPin, gpio_output, gpio_nopull);
+	_nrf91_gpioConfig(console_common.rxPin, gpio_input, gpio_nopull);
+	_nrf91_gpioConfig(console_common.rtsPin, gpio_output, gpio_nopull);
+	_nrf91_gpioConfig(console_common.ctsPin, gpio_input, gpio_pulldown);
+
+	_nrf91_gpioSet(console_common.txPin, gpio_high);
+	_nrf91_gpioSet(console_common.rtsPin, gpio_high);
+}
+
+
+void _hal_consolePrint(const char *s)
+{
+	volatile char *tx_dma_buff = (volatile char *)console_common.txDma;
+	int iter, len, blocks, pos, cnt = 0;
+
+	len = hal_strlen(s);
+	blocks = len / TX_DMA_SIZE;
+
+	pos = 0;
+	for (iter = 0; iter <= blocks; iter++) {
+		cnt = 0;
+		while ((s[pos] != '\0') && (cnt < console_common.txDmaSize)) {
+			tx_dma_buff[cnt++] = s[pos++];
+		}
```

Comment:
I've verified that this solution with using s is not working

Reply:
Ok, that's a shame though, we could save both buffer space on copy operation :(

I don't quite understand though why `console_common.txDma` is working. There should be nothing special about this memory in .bss.

Reply:
What do you mean? Why bss is stored above `0x2000 0000` (RAM) and strings not?

Reply:
from what I understood the string originates from `.rodata` in that case and resides on flash. the EasyDMA can't access directly the FLASH memory so we need to copy it to ram (https://infocenter.nordicsemi.com/pdf/nRF9160_PS_v2.0.pdf page 22)

Please note that on Zephyr the copying is done **only if the original string is not in RAM** so that would be the optimal solution imho.

Reply:
Yeah, that's right. I proposed sth similar in the last force push.